src/ucp/tag/rndv.c

```diff
@@ -610,6 +617,228 @@ static void ucp_rndv_req_send_rma_get(ucp_request_t *rndv_req, ucp_request_t *rr
+                      uct_completion_t *self, ucs_status_t status)
+{
+    ucp_request_t *freq     = ucs_container_of(self, ucp_request_t,
+                                               send.state.uct_comp);
+    ucp_request_t *rndv_req = freq->send.rndv_get.rreq;
+    ucp_request_t *rreq     = rndv_req->send.rndv_get.rreq;
+
+    ucs_trace_req("freq:%p: recv_frag_get done. rreq:%p length:%ld offset:%ld",
+                  freq, rndv_req, freq->send.length,
+                  freq->send.rndv_get.remote_address - rndv_req->send.rndv_get.remote_address);
+
+    ucp_rndv_recv_frag_put_mem_type(rreq, rndv_req, freq,
+                                    (ucp_mem_desc_t *)freq->send.buffer -1,
+                                    freq->send.length, (freq->send.rndv_get.remote_address -
+                                    rndv_req->send.rndv_get.remote_address));
+}
+
+static ucs_status_t
+ucp_rndv_recv_start_get_pipeline(ucp_worker_h worker, ucp_request_t *rndv_req,
+                                 ucp_request_t *rreq, uintptr_t remote_request,
+                                 const void *rkey_buffer, uint64_t remote_address,
+                                 size_t size, size_t base_offset)
+{
+    ucp_context_h context = worker->context;
+    ucp_mem_desc_t *mdesc;
+    ucp_request_t *freq;
+    ucs_status_t status;
+    size_t max_frag_size, offset, length;
+
+    max_frag_size                          = context->config.ext.rndv_frag_size;
+    rndv_req->send.rndv_get.remote_request = remote_request;
+    rndv_req->send.rndv_get.remote_address = remote_address - base_offset;
+    rndv_req->send.rndv_get.rreq           = rreq;
+    rndv_req->send.length                  = size;
+    rndv_req->send.state.dt.offset         = 0;
+
+    /* Protocol:
+     * Step 1: GET remote fragment into HOST fragment buffer
+     * Step 2: PUT from fragment buffer to MEM TYPE destination
+     * Step 3: Send ATS for RNDV request
+    */
+
+    status = ucp_ep_rkey_unpack(rndv_req->send.ep, rkey_buffer,
+                                &rndv_req->send.rndv_get.rkey);
+    if (ucs_unlikely(status != UCS_OK)) {
+        ucs_fatal("failed to unpack rendezvous remote key received from %s: %s",
+                  ucp_ep_peer_name(rndv_req->send.ep), ucs_status_string(status));
+    }
+
+    offset = 0;
+    while (offset != size) {
+        length = ucs_min(size - offset, max_frag_size);
+
+        freq = ucp_request_get(worker);
+        if (ucs_unlikely(freq == NULL)) {
+            ucs_error("failed to allocate fragment receive request");
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        mdesc = ucp_worker_mpool_get(&worker->rndv_frag_mp);
+        if (ucs_unlikely(mdesc == NULL)) {
+            ucs_error("failed to allocate fragment memory desc");
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        /* GET remote fragment into HOST fragment buffer */
+        freq->send.ep                      = rndv_req->send.ep;
```

Comment:
looks very similar to `ucp_rndv_send_frag_get_mem_type` except ep type.
Can it be united as well somehow?

Reply:
unified some code.  can you please review